File: sdk/go/hash.go
package experimentation

import (
	"crypto/md5" //nolint:gosec // MD5 used for consistent hashing, not cryptographic security
	"encoding/binary"
)

// hashDivisor is 2^32 (0x100000000), used to normalize the MD5-derived
// uint32 to a float64 in [0.0, 1.0). It matches every other platform SDK and
// the backend's consistent_hash module.
const hashDivisor = float64(0x100000000)

// ConsistentHash returns a deterministic float64 in [0.0, 1.0) for the given
// user ID and key (flag or experiment key).
//
// Algorithm (byte-for-byte compatible with the other platform SDKs and the
// golden vectors in tests/sdk-contract/golden-vectors.json):
//  1. Concatenate userID + ":" + key as UTF-8.
//  2. Compute the MD5 digest.
//  3. Interpret the first 4 bytes as a little-endian unsigned 32-bit integer.
//  4. Divide by 2^32.
//
// The SDK does not use this to decide variants any more (the server does);
// it is exported so integrations can verify cross-SDK parity.
func ConsistentHash(userID, key string) float64 {
	input := userID + ":" + key
	digest := md5.Sum([]byte(input)) //nolint:gosec
	v := binary.LittleEndian.Uint32(digest[:4])
	return float64(v) / hashDivisor
}

File: sdk/js/src/hash.ts
import { md5Bytes } from './md5';

/** 2^32 — the divisor shared by every SDK (NOT 2^32 - 1). */
const HASH_DIVISOR = 4294967296;

/** Lower-case hex MD5 digest of the UTF-8 encoding of `input`. */
export function md5Hex(input: string): string {
  return Array.from(md5Bytes(input))
    .map(b => b.toString(16).padStart(2, '0'))
    .join('');
}

/**
 * Cross-SDK consistent hash in `[0, 1)`:
 * `MD5("{userId}:{flagKey}")` → first 4 bytes as little-endian uint32 → ÷ 2^32.
 *
 * Exported as a utility only. The server decides assignments and flag
 * evaluations; nothing in this SDK uses the hash to pick a variant.
 */
export function consistentHash(userId: string, flagKey: string): number {
  const digest = md5Bytes(`${userId}:${flagKey}`);
  const uint32 = new DataView(digest.buffer).getUint32(0, true);
  return uint32 / HASH_DIVISOR;
}

export { md5Bytes };
